toxcore/group.c

```diff
@@ -3392,116 +3392,166 @@ uint8_t *conferences_save(const Group_Chats *g_c, uint8_t *data)
-            }
+        // NOTE: this relies on friends being loaded before conferences.
+        peer->is_friend = (getfriend_id(g_c->m, peer->real_pk) != -1);
+    }
+
+    if (g->numfrozen > g->maxfrozen) {
+        g->maxfrozen = g->numfrozen;
+    }
 
-            Group_Peer *peer = &g->frozen[j];
-            memset(peer, 0, sizeof(Group_Peer));
+    g->status = GROUPCHAT_STATUS_CONNECTED;
 
-            id_copy(peer->real_pk, data);
-            data += CRYPTO_PUBLIC_KEY_SIZE;
-            id_copy(peer->temp_pk, data);
-            data += CRYPTO_PUBLIC_KEY_SIZE;
+    id_copy(g->real_pk, nc_get_self_public_key(g_c->m->net_crypto));
 
-            lendian_bytes_to_host16(&peer->peer_number, data);
-            data += sizeof(uint16_t);
+    return (uint32_t)(data - init_data);
+}
 
-            lendian_bytes_to_host64(&peer->last_active, data);
-            data += sizeof(uint64_t);
+non_null()
+static State_Load_Status load_conferences(Group_Chats *g_c, const uint8_t *data, uint32_t length)
+{
+    const uint8_t *init_data = data;
+    uint32_t num_groups;
 
-            peer->nick_len = *data;
+    while (length >= (uint32_t)(data - init_data) + SAVED_CONF_SIZE_CONSTANT) {
+        const int groupnumber = create_group_chat(g_c);
 
-            if (peer->nick_len > MAX_NAME_LENGTH) {
-                return STATE_LOAD_STATUS_ERROR;
-            }
+        if (groupnumber == -1) {
+            // If this fails there's a serious problem, don't bother with cleanup
+            return STATE_LOAD_STATUS_ERROR;
+        }
 
-            ++data;
+        Group_c *g = &g_c->chats[groupnumber];
 
-            if (length < (uint32_t)(data - init_data) + peer->nick_len) {
-                return STATE_LOAD_STATUS_ERROR;
-            }
+        uint32_t consumed = load_group(g, g_c, data, length - (uint32_t)(data - init_data));
 
-            memcpy(peer->nick, data, peer->nick_len);
-            data += peer->nick_len;
+        if (consumed == 0) {
+            // remove partially loaded stuff
+            bool ret = wipe_group_chat(g_c, groupnumber);
 
-            // NOTE: this relies on friends being loaded before conferences.
-            peer->is_friend = (getfriend_id(g_c->m, peer->real_pk) != -1);
+            // wipe_group_chat must be able to wipe a partially loaded group
+            assert(ret == true);
+            goto LOAD_ERROR;
         }
 
-        if (g->numfrozen > g->maxfrozen) {
-            g->maxfrozen = g->numfrozen;
-        }
+        data += consumed;
 
-        g->status = GROUPCHAT_STATUS_CONNECTED;
-        memcpy(g->real_pk, nc_get_self_public_key(g_c->m->net_crypto), CRYPTO_PUBLIC_KEY_SIZE);
         const int peer_index = addpeer(g_c, groupnumber, g->real_pk, dht_get_self_public_key(g_c->m->dht), g->peer_number,
                                        nullptr, true, false);
 
         if (peer_index == -1) {
-            return STATE_LOAD_STATUS_ERROR;
+            goto LOAD_ERROR;
```

Comment:
Let's try to avoid adding more `goto` and instead use an inner function with a cleanup outside.

Reply:
Changed the structure, though no inner function.

Reply:
Finally understand how you meant that and did it.